transports/tcp/src/lib.rs

```diff
@@ -178,16 +174,187 @@ impl Transport for TcpConfig {
+
+            let local_addr = match sock.local_addr() {
+                Ok(sock_addr) => {
+                    if let Addresses::Many(ref mut addrs) = self.addrs {
+                        if let Err(err) = check_for_interface_changes(&sock_addr, self.port, addrs, &mut self.pending) {
+                            return (Err(err), self);
+                        }
+                    }
+                    ip_to_multiaddr(sock_addr.ip(), sock_addr.port())
+                }
+                Err(err) => {
+                    debug!("Failed to get local address of incoming socket: {:?}", err);
+                    continue
+                }
+            };
+
+            let remote_addr = ip_to_multiaddr(sock_addr.ip(), sock_addr.port());
+
+            match $apply_config(&self.config, &sock) {
+                Ok(()) => {
+                    trace!("Incoming connection from {} at {}", remote_addr, local_addr);
+                    self.pending.push_back(Ok(ListenerEvent::Upgrade {
+                        upgrade: future::ok($tcp_trans_stream { inner: sock }),
+                        local_addr,
+                        remote_addr
+                    }))
+                }
+                Err(err) => {
+                    debug!("Error upgrading incoming connection from {}: {:?}", remote_addr, err);
+                    self.pending.push_back(Ok(ListenerEvent::Upgrade {
+                        upgrade: future::err(err),
+                        local_addr,
+                        remote_addr
+                    }))
+                }
+            }
+        }
+    }
+}
+
+/// Wraps around a `TcpStream` and adds logging for important events.
+#[cfg_attr(docsrs, doc(cfg(feature = $feature_name)))]
+#[derive(Debug)]
+pub struct $tcp_trans_stream {
+    inner: $tcp_stream,
+}
+
+impl Drop for $tcp_trans_stream {
+    fn drop(&mut self) {
+        if let Ok(addr) = self.inner.peer_addr() {
+            debug!("Dropped TCP connection to {:?}", addr);
+        } else {
+            debug!("Dropped TCP connection to undeterminate peer");
+        }
+    }
+}
+
+/// Applies the socket configuration parameters to a socket.
+fn $apply_config(config: &$tcp_config, socket: &$tcp_stream) -> Result<(), io::Error> {
+    if let Some(ttl) = config.ttl {
+        socket.set_ttl(ttl)?;
+    }
+
+    if let Some(nodelay) = config.nodelay {
+        socket.set_nodelay(nodelay)?;
+    }
+
+    Ok(())
+}
+
+};
+}
+
+#[cfg(feature = "async-std")]
+codegen!("async-std", TcpConfig, TcpTransStream, TcpListenStream, apply_config_async_std, async_std::net::TcpStream, async_std::net::TcpListener);
+
+#[cfg(feature = "tokio")]
+codegen!("tokio", TokioTcpConfig, TokioTcpTransStream, TokioTcpListenStream, apply_config_tokio, tokio::net::TcpStream, tokio::net::TcpListener);
+
+#[cfg(feature = "async-std")]
+impl AsyncRead for TcpTransStream {
+    fn poll_read(mut self: Pin<&mut Self>, cx: &mut Context, buf: &mut [u8]) -> Poll<Result<usize, io::Error>> {
```

Comment:
Why not include `AsyncRead` and `AsyncWrite` into `code_gen` as well? Aren't they pretty much identical apart from their import paths?

```diff
diff --git a/async-std.rs b/tokio.rs
index 260f1b6..bb0b995 100644
--- a/async-std.rs
+++ b/tokio.rs
@@ -1,21 +1,21 @@
-#[cfg(feature = "async-std")]
-impl AsyncRead for TcpTransStream {
+#[cfg(feature = "tokio")]
+impl AsyncRead for TokioTcpTransStream {
     fn poll_read(mut self: Pin<&mut Self>, cx: &mut Context, buf: &mut [u8]) -> Poll<Result<usize, io::Error>> {
-        AsyncRead::poll_read(Pin::new(&mut self.inner), cx, buf)
+        tokio::io::AsyncRead::poll_read(Pin::new(&mut self.inner), cx, buf)
     }
 }
 
-#[cfg(feature = "async-std")]
-impl AsyncWrite for TcpTransStream {
+#[cfg(feature = "tokio")]
+impl AsyncWrite for TokioTcpTransStream {
     fn poll_write(mut self: Pin<&mut Self>, cx: &mut Context, buf: &[u8]) -> Poll<Result<usize, io::Error>> {
-        AsyncWrite::poll_write(Pin::new(&mut self.inner), cx, buf)
+        tokio::io::AsyncWrite::poll_write(Pin::new(&mut self.inner), cx, buf)
     }
 
     fn poll_flush(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Result<(), io::Error>> {
-        AsyncWrite::poll_flush(Pin::new(&mut self.inner), cx)
+        tokio::io::AsyncWrite::poll_flush(Pin::new(&mut self.inner), cx)
     }
 
     fn poll_close(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Result<(), io::Error>> {
-        AsyncWrite::poll_close(Pin::new(&mut self.inner), cx)
+        tokio::io::AsyncWrite::poll_shutdown(Pin::new(&mut self.inner), cx)
     }
 }
```

Reply:
I thought it would start being too confusing to have to pass 4 additional method names to the macro.